cambalache: init at 0.8.0

jtojnar opened the pull request:
###### Motivation for this change
UI designer for GTK 3 & 4 (sort of like Glade):

https://gitlab.gnome.org/jpu/cambalache

- [x] GTK 3 works but GTK 4 currently does not and no errors are logged, needs debugging. https://gitlab.gnome.org/jpu/cambalache/-/issues/28
- [x] MIME declarations (`$out/share/mime/`) are required for filters in „Open project“ dialogue, can we add them to XDG_DATA_DIRS or something – turns out it is a bug https://gitlab.gnome.org/jpu/cambalache/-/merge_requests/7
- [x] Opening projects is broken on main branch https://gitlab.gnome.org/jpu/cambalache/-/issues/29
- [x] GTK 4’s Broadway daemon should be moved out of `dev` output to reduce runtime closure.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jtojnar committed:
cambalache: init at 0.8.2

r-rmcgibbo commented:
Result of `nixpkgs-review pr 138940` at a9001915 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>cambalache</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `stdenv`.
Near pkgs/development/tools/cambalache/default.nix:1:3:
```
  |
1 | { stdenv
  |   ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/tools/cambalache/default.nix:
```diff
@@ -0,0 +1,93 @@
+    wrapGAppsHook
+  ];
+
+  pythonPath = [
+    python3.pkgs.pygobject3
+    python3.pkgs.lxml
+  ];
```
```suggestion
  pythonPath = with python3.pkgs; [
    pygobject3
    lxml
  ];
```

SuperSandro2000 commented on pkgs/development/tools/cambalache/default.nix:
```diff
@@ -0,0 +1,93 @@
+    webkitgtk
+  ];
+
+  # Not compatible with g-i setup hooks.
```
```suggestion
  # Not compatible with gobject-introspection setup hooks.
```

SuperSandro2000 commented on pkgs/development/tools/cambalache/default.nix:
```diff
@@ -0,0 +1,95 @@
+    # Let python wrapper use GNOME flags.
+    makeWrapperArgs+=(
+      # For gtk4-broadwayd
+      --prefix PATH : "${gtk4.dev}/bin"
```
Should we use lib.getBin here?

jtojnar replied:
`lib.getBin` does not respect `outputBin` attribute.

psydvl commented:
Newbie questions:
1. Why python3.pkgs instead of python3Packages?
2. Why unstableGitUpdater, nor stable?
3. And why is it still draft? I successfully built v0.8.0 (latest) with similar build file...

jtojnar commented:
> * Why python3.pkgs instead of python3Packages?

It is shorter. And these days, the latter is alias of the former.
 
> * Why unstableGitUpdater, nor stable?

I had to switch to git revision because 0.7.4 was broken. Now that 0.8.0 has been released, that probably will no longer be necessary.

> * And why is it still draft? I successfully built v0.8.0 (latest) with similar build file...

See the checkboxes in the OP.

psydvl commented:
Idea:
Since gtk4-broadwayd not necessary dependency and for now not separable from gtk4.dev, it can be brought out as option like `useGTK4Broadwayd`

``` patch
-{ stdenv
+{ stdenv, lib
---
---
+, useGTK3Broadwayd ? false
+, useGTK4Broadwayd ? false
}:
```

``` patch
  preFixup = ''
    # Let python wrapper use GNOME flags.
    makeWrapperArgs+=(
+      # For gtk3 broadwayd
+      ${ lib.optionalString useGTK3Broadwayd ''--prefix PATH : "${gtk3.dev}/bin"''}
      # For gtk4-broadwayd
-      --prefix PATH : "${gtk4.dev}/bin"
+      ${ lib.optionalString useGTK3Broadwayd ''--prefix PATH : "${gtk4.dev}/bin"''}
      "''${gappsWrapperArgs[@]}"
    )
  '';
```

Then app can be used as `pkgs.cambalache` or 
```nix
pkgs.cambalache.override {
  # useGTK3Broadwayd = true;
  useGTK4Broadwayd = true;
}
``` 
Also, there are can be `cambalache-dev` added to `all-packages.nix` or dev (`cambalache.dev`) output created

jtojnar commented:
Broadway is necessary dependency – it is used for rendering the designed interface. And I think it should not be that hard – just move `broadwayd` (only on GTK 4 by the way, GTK3 does not have a separate binary) to a new output in `postInstall`. I just have not gotten to doing it yet.

The path to the `gtk4.dev` is baked into the program using a wrapper so adding `cambalache.dev` output would do nothing.

jtojnar committed:
gtk4: move broadwayd into out

It is convenient to have it in PATH by default and does not seem to increase the output size much (+0.2M).

Just as was done to GTK 3 in https://github.com/NixOS/nixpkgs/pull/116053

jtojnar commented:
Thanks, I did not realize we did https://github.com/NixOS/nixpkgs/pull/116053 so I thought GTK 3 did not use the daemon. I decided to do the same in GTK 4 since it is quite small.

jtojnar commented:
Not sure if it is makes sense to merge this since it is still not very usable at the moment.

psydvl commented:
Since `pythonPath` is deprecated https://github.com/NixOS/nixpkgs/blob/e0f466e9ffef37e2d97e6135df17814fe4fe82ca/pkgs/development/interpreters/python/mk-python-derivation.nix#L46-L47
Shouldn't it be replaced with `propagatedBuildInputs`?

jtojnar commented:
Hopefully it will not be deprecated for long: https://github.com/NixOS/nixpkgs/pull/102613

psydvl commented:
There are 0.8.2 already published
https://gitlab.gnome.org/jpu/cambalache/-/tags
And it's so usable, so glade is
Why not to merge?

noonien commented:
Any update on this?
